bundles/remote_services/remote_service_admin_dfi/src/remote_service_admin_dfi_constants.h

```diff
@@ -36,6 +36,15 @@
 #define RSA_DFI_CONFIGURATION_TYPE      "org.amdatu.remote.admin.http"
 #define RSA_DFI_ENDPOINT_URL            "org.amdatu.remote.admin.http.url"
 
+/**
+ * @brief RSA Configuration type for zeroconf http, it is synonymous(https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.remoteservices.html#i1698916) with RSA_DFI_CONFIGURATION_TYPE, they refer to the same endpoint.
```

Comment:
I understand this different.

As I understand it, the service.exported.configs and service.imported.configs refer to remote service admin configurations and not discovery. 
Different discovery mechanism should work without knowledge of the remote service admins used and vica versa. 

For example 
-  RSA DFI adds a "org.amdatu.remote.admin.http.url" property when creating a endpoint for an export service.   
- Discovery (Zeroconf, configured or etcd) "just" announces and discovers the endpoints. 
- RSA DFI checks the config when importing a service through a endpoint and uses the "org.amdatu.remote.admin.http.url".

It is possible to support configuration and endpoint with a "ifname", "port", "ipdresses" and "path" next to a "url" property, but IMO this is then an extension to RSA DFI and not a ZEROCONF concern.

For example (snippet):

```C

if (strncmp(configType, RSA_DFI_CONFIGURATION_TYPE, 1024) == 0) {
    const char *url = celix_properties_get(endpoint->properties, RSA_DFI_ENDPOINT_URL, NULL);
    if (url) {
        return true; //RSA_DFI config with  endpoint url
   }
       
    const char *ip = celix_properties_get(endpoint->properties, CELIX_RSA_DFI_ZEROCONF_ENDPOINT_IPADDRESSES, NULL);//the property is set by the discovery_zeroconf
    const char *port = celix_properties_get(endpoint->properties, CELIX_RSA_DFI_ZEROCONF_ENDPOINT_PORT, NULL);
    const char *path = celix_properties_get(endpoint->properties, CELIX_RSA_DFI_ZEROCONF_ENDPOINT_PATH, NULL);
        if (ip != NULL || port != NULL || path != NULL) {
            return true; //RSA_DFI config with a endpoint ip, port and path/
        }
} 
return false;
```

Reply:
Here are some inconsistencies between RSA and ZEROCONF: RSA may not bind all (i.e. not bind to a specific ip address) while ZEROCONF will advertise all addresses associated with a network interface.

Reply:
> IMO the zeroconf discovery and remote service admins are now entangled and this is not needed.

Adding the host name resolution mechanism to `zeroconf_discovery` does cause `RSA` to become entangled with `zeroconf_discovery`.
The main reason for adding the host name resolution mechanism to `zeroconf_discovery` is that if `RSA` binds to any address (0.0.0.0), then it does not need to configure a specific IP, and `zeroconf_discovery` resolves  IP list  from the host name, then gives the IP list to the client via the " <config_type>.ipaddresses" endpoint property. It means that if we use `zeroconf_discovery`, the remote service will still work even if `RSA` does not configure an IP.

For the above issue, should we do host name resolution in `zeroconf_discovery`?
If we use host name in the URL of the endpoint (e.g. http://computer.local:8080/service/example) maybe that will resolve the problem as well. (Host name resolution will be done by the `RSA`)

If we do host name resolution in `zeroconf_discovery`, we should design interfaces between `RSA` and `zeroconf_discovery` for distinguishing whether host name resolution is needed or not, because we don't need to do host name resolution for an endpoint that already specifies a specific IP. In my option, we can add a specific property(e.g. `<config_type>.automatically_get_ip_by_discovery`) to the endpoint description, and `zeroconf_discovery` will do host name resolution according to this property.

Reply:
>  It means that if we use zeroconf_discovery, the remote service will still work even if RSA does not configure an IP.

That is a nice feature and for now I cannot think of a way to this without some coupling between the RSA and discovery.

The option is what you mention is a good way to opt-in for discovery specific behaviour. 
Maybe more generic and not coupled to zeroconf. e.g:
 - If a remote endpoint has property has a property "CELIX_RSA_DISCOVERY_IP_ADRESSES" then the value will be updated to discovered ip addresses.

 
If I look at the RSA spec (https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.remoteserviceadmin.html#org.osgi.service.remoteserviceadmin.RemoteServiceAdmin), `exportService` does allow to return multiple ExportRegistration and thus multiple endpoints. 

Maybe this can be combined and the RSA DFI can provide mutiple ExportRegistrations. 1 for the fixed IP and 1 for a dynamic IP adresses. 

I will read the spec a bit more and check whether exporting multiple ExportRegistration (with multiple paths to the same exported service) is feasible.

Reply:
I have some new thoughts about the dynamic ip assignment resolving challenges. 
First off, I think is good to come to a sound conclusion, but this does not mean that this should be directly solved in this issue. 

So if we agree on a desired solution, I will make a ticket for this and we can then decide what should flow back to this PR.

My thoughts:
a) Let RSA implementations decide if they support "dynamic IP fill-in"  by providing a additional service property on the registered RSA service
b) Optionally RSA export registration that support dynamic IP fill-in should also opt-int for dynamic replacement by adding a endpoint property to the export registration properties
c) Let discovery implementations decide if the they support "interface specific" endpoints by providing a additional service property on the registered endpoint listener interface.
c) Update the Topology Manager to detect RSA services what support dynamic IP fill-in and in that case create multiple endpoints for a single export registration based on the available network interfaces and an optional network selection configuration. These endpoints are then forwarded to the discovery listener services that support network specific endpoints. The discovery implementation can pick and handle or drop endpoints based on the selected network interface. 

The primary idea is that most of the boiler-plating to support dynamic IP fill-in is moved to the topology manager and kept away from the RSA and discovery implementation.

For the service/endpoint property something like the following can be used:
- RSA Dynamic IP fill-in support
  - Property name: "rsa.dynamic.ip.support"
  - Type boolean
  - Indicates whether the RSA implementation supports dynamic IP address fill-in for service exports
- RSA Export Registration replacement opt-in
  - Property name: rsa.export.dynamic.ip.optin
  - Type boolean
  - Description: Signals if a service export opts in for dynamic IP address replacement 
- RSA Export Registration dynamic replaceable property
  - Property name: rsa.ip.address
  - Type string
  - Description: The dynamically determined IP address for the endpoint.
 - Discovery interface-specific endpoints support.
   - Property Name: "discovery.interface.specific.endpoints.support
   - Type boolean
   - Description:  Indicates if the discovery implementation supports handling of interface-specific endpoints.
 

I understand that this is different from the current solution, but maybe a more cleaner approach.

WDYT?

Reply:
> Thank you for your reply.
> 
> First, I agree with this solution. But I still have some issues that need to confirm with you.
> 
> > Update the Topology Manager to detect RSA services what support dynamic IP fill-in and in that case create multiple endpoints for a single export registration based on the available network interfaces and an optional network selection configuration. These endpoints are then forwarded to the discovery listener services that support network specific endpoints. The discovery implementation can pick and handle or drop endpoints based on the selected network interface.
> 
> Does this mean that the endpoint descriptions for dynamic IP are generated by Topology Manager when exporting service? Should the network interface used to publish services be specified by Topology Manager or RSA?

Good point, this should indeed be the RSA. 

> 
> How about dealing with it in two cases: In the first case, if the RSA supports dynamic IP, RSA should bind its services to ANY address (0.0.0.0). Then the Topology Manager should specify which network interfaces the services are published to, and restrict access to the corresponding services from other network interfaces (e.g.: `iptables -A INPUT -i eth0 -p tcp -- dport 80 -j DROP`).
> 
> In the second case, if RSA does not support dynamic IP, RSA binds its network services to a specific IP, and its exported endpoint should specify which network interface the service is published to.
> 
> In the above case, we will define a generic endpoint property `rsa.ifnames` (Type string list). The property is used to specify which network interfaces the service is published to.

That does indeed sound better. One of my cancern was to prevent too much repeated boiler plaiting between the RSA bundles, but indeed the RSA is binding on network interface and hence it should decide and informs whether is uses a single or multiple interfaces. 
> 
> > RSA Export Registration dynamic replaceable property
> > 
> > * Property name: rsa.ip.address
> > * Type string
> 
> When using dynamic IP, perhaps multiple IP are obtained, should the `rsa.ip.address` type be a string list?

List of string, but this is not yet supported. I am busy with a branch that updates properties for (among others) list of strings: https://github.com/apache/celix/tree/feature/674-improve-properties

I will make a ticket for yet another improvement on remote service and let it depend on #674 

For this PR this remark can now be ignored.